vita3k/modules/SceHttp/SceHttp.cpp

```diff
@@ -145,57 +504,454 @@ EXPORT(int, sceHttpGetIcmOption) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceHttpGetLastErrno) {
-    return UNIMPLEMENTED();
+EXPORT(int, sceHttpGetLastErrno, int reqId, int *errNum) {
+    if (!host.http.inited)
+        return RET_ERROR(SCE_HTTP_ERROR_BEFORE_INIT);
+
+    if (host.http.requests.find(reqId) == host.http.requests.end())
+        return RET_ERROR(SCE_HTTP_ERROR_INVALID_ID);
+
+    *errNum = (int)errno;
+
+    return 0;
 }
 
-EXPORT(int, sceHttpGetMemoryPoolStats) {
+EXPORT(int, sceHttpGetMemoryPoolStats, SceHttpMemoryPoolStats *currentStat) {
+    if (!host.http.inited)
+        return RET_ERROR(SCE_HTTP_ERROR_BEFORE_INIT);
+
+    if (!currentStat)
+        return RET_ERROR(SCE_HTTP_ERROR_INVALID_VALUE);
+
     return UNIMPLEMENTED();
 }
 
 EXPORT(int, sceHttpGetNonblock) {
     return UNIMPLEMENTED();
 }
 
-EXPORT(int, sceHttpGetResponseContentLength) {
-    return UNIMPLEMENTED();
-}
+EXPORT(int, sceHttpGetResponseContentLength, int reqId, size_t *contentLength) {
+    if (!host.http.inited)
+        return RET_ERROR(SCE_HTTP_ERROR_BEFORE_INIT);
 
-EXPORT(int, sceHttpGetStatusCode) {
-    return UNIMPLEMENTED();
-}
+    if (host.http.requests.find(reqId) == host.http.requests.end())
+        return RET_ERROR(SCE_HTTP_ERROR_INVALID_ID);
+
+    auto req = host.http.requests.find(reqId);
+
+    char contLen[15] = "Content-Length";
+
+    auto length_it = req->second.res.headers.find(contLen);
+    if (length_it == req->second.res.headers.end())
+        return RET_ERROR(SCE_HTTP_ERROR_NO_CONTENT_LENGTH);
 
-EXPORT(int, sceHttpInit) {
-    STUBBED("Fail when called");
-    return -1;
+    auto len = length_it->second;
+    int length = std::stoi(len);
+    *contentLength = length;
+
+    req->second.res.contentLength = length;
+
+    return 0;
 }
 
-EXPORT(int, sceHttpParseResponseHeader) {
-    return UNIMPLEMENTED();
+EXPORT(int, sceHttpGetStatusCode, int reqId, int *statusCode) {
+    if (!host.http.inited)
+        return RET_ERROR(SCE_HTTP_ERROR_BEFORE_INIT);
+
+    if (host.http.requests.find(reqId) == host.http.requests.end())
+        return RET_ERROR(SCE_HTTP_ERROR_INVALID_ID);
+
+    auto req = host.http.requests.find(reqId)->second;
+
+    *statusCode = req.res.statusCode;
+    return 0;
 }
 
-EXPORT(int, sceHttpParseStatusLine) {
-    return UNIMPLEMENTED();
+EXPORT(int, sceHttpInit, size_t poolSize) {
+    if (host.http.inited)
+        return RET_ERROR(SCE_HTTP_ERROR_ALREADY_INITED);
+
+    if (poolSize == 0)
+        return RET_ERROR(SCE_HTTP_ERROR_INVALID_VALUE);
+
+    STUBBED("ignore poolSize");
+
+    if (host.http.sslInited)
+        host.http.ssl_ctx = SSL_CTX_new(SSLv23_client_method());
```

Comment:
Games call HttpInit and then SslInit so this line will never be called, it should be moved to sslInit.

Reply:
VitaShell calls SSL then HTTP inits, probably there is no standard on which order they get called, will improve

Reply:
Oh, alright.